File: back-ends/nodejs-rest/app/controllers/posts-controller.js
'use strict';

import Post from './../models/Post';
import { generalErrorResponse, permissionsErrorResponse, unauthorizedErrorResponse } from '../utils/error-factory';

// Create a new post
module.exports.create = (req, res) => {
  if (!req.currentUser) {
    return unauthorizedErrorResponse(res);
  }

  let path = req.body.post.title.replace(/\s/g, '-') + Date.now();

  return Post.create({
    url_path: path,
    title: req.body.post.title,
    body: req.body.post.body,
    _author: req.currentUser._id,
    tags: [],
  })
  .then(post => {
    return res.json({
      post: post,
    });
  })
  .catch((err) => {
    return generalErrorResponse(res);
  });
};

// View a list of all posts
module.exports.index = (req, res) => {
  if(req.query.url_path) {
    return Post.findOne({
      'url_path': req.query.url_path,
    }).populate('tags')
    .then(post => {
      if(post) {
        return res.json({
          post: post,
        });
      } else {
        return res.sendStatus(404);
      }
    });
  } else {
    return Post.find()
    .populate('tags')
    .then(posts => {
      return res.json({
        posts: posts,
      });
    })
    .catch((err) => {
      return generalErrorResponse(res);
    });
  }
};

// Read a post
module.exports.show = (req, res) => {
  return Post.findById(req.params.id)
  .populate('tags')
  .then(post => {
    if(post) {
      return res.json({
        post: post,
      });
    } else {
      res.sendStatus(404);
    }
  })
  .catch((err) => {
    return generalErrorResponse(res);
  });
};

// Update a post
module.exports.update = (req, res) => {
  if (!req.currentUser) {
    return unauthorizedErrorResponse(res);
  }

  return Post.findById(req.params.id)
  .populate('tags')
  .then(post => {
    if (post._author.toString() === req.currentUser._id.toString()) {
      let path = req.body.post.url_path.replace(/\s/g, '-');

      post.url_path = path || post.url_path;
      post.title = req.body.post.title || post.title;
      post.body = req.body.post.body || post.body;
      post._author = post.user_id;

      return post.save()
      .then(post => {
        return res.json({
          post: post,
        });
      });
    } else {
      return permissionsErrorResponse(res);
    }
  })
  .catch((err) => {
    return generalErrorResponse(res);
  });
};

// Delete a post
module.exports.delete = (req, res) => {
  if (!req.currentUser) {
    return unauthorizedErrorResponse(res);
  }

  return Post.findById(req.params.id)
  .then(post => {
    if(String(post._author) === String(req.currentUser._id)) {
      return post.remove();
    } else {
      return permissionsErrorResponse(res);
    }
  })
  .then(() => {
    return res.json({
      deleted_id: req.params.id,
    });
  })
  .catch((err) => {
    return generalErrorResponse(res);
  });
};

module.exports.getPostsByTag = (req, res) => {
  return Post.find({'tags': req.params.tag_id})
  .populate('tags')
  .then(posts => {
    return res.json({
      posts: posts,
    });
  })
  .catch((err) => {
    return generalErrorResponse(res);
  });
};

module.exports.postsByUser = (req, res) => {
  return Post.find({
    _author: req.params.user_id,
  })
  .then(list => {
    return res.json({
      posts: list,
    });
  })
  .catch((err) => {
    return generalErrorResponse(res);
  });
};

File: back-ends/rails-rest/app/controllers/posts_controller.rb
class PostsController < BaseController
  before_action :set_user
  before_filter :ensure_post_exists, only: :show
  before_filter :check_url_path, only: :create
  before_filter :ensure_user_exists, only: :get_post_by_user
  skip_before_filter :authenticate_user_from_token!, :except => [:create, :update, :delete]

  # @description GET /posts   
  # @return {Object} posts - JSON Object of the Post 
  # @param {String} url_path - Searches for a particular post with url_path
  # @param {String} Authorization - Authentication token. Format is'Bearer: tokenvalue' and is required.
  def index
    if !params[:url_path].blank?
      @post = Post.where(:url_path => params[:url_path])
      render 'posts/show'
    else 
      @posts = Post.all 
      'posts/index'
    end 
  end

  # @description POST /posts   
  # @return {Object} post - JSON Object of the Post 
  # @param {String} title - The Title of the Post. Required.
  # @param {String} body - The Body of the Post. Required.
  def create
    @post = @user.posts.create(post_params)
    @post.url_path = @url_path 

    # Successful JSON in /views/posts/create.json.rabl

    # On Error /views/helpers/error
    respond_with @post do 
      'helpers/errors' unless @post.save
    end
  end

  # @description GET /posts/:id   
  # @return {Object} post - JSON Object of the Post 
  def show
    @post = Post.find(params[:id]) 
  end

  # @description GET /tags/:tag_id/posts 
  # @return {Object} post - JSON Object of the Post 
  def get_post_by_tag
    # TO DO -> Make after 
  end

  # @description GET /users/:user_id/posts
  # @return {Object} posts - JSON Object of the Post 
  # @param {String} user_id - The user's id  
  def get_post_by_user
    puts params[:user_id]
    @posts = Post.where(:user_id => params[:user_id])
    render 'posts/index'
  end

  # @description PUT /users/:id
  # @param {Int} id - User's id. Implicit Paramater. Is required.  
  # @param {String} Authorization - Authentication token. Format is'Bearer: tokenvalue' and is required.
  #      
  # @returns JSON Object of User 
  # @throws JSON Object of Errors
  def update
    @post = Post.find(params[:id])

    updates = post_params.as_json
  
    if @post.update(updates)
      render 'posts/show'
    else
      respond_with @post do 
        'helpers/errors'
      end
    end
  end

  # @description DELETE /posts/:id 
  # @param {string} id - Post's id. Implicit Paramater. Is required.  
  # @param {string} Authorization - Authentication token. Format is'Bearer: tokenvalue' and is required.
  #      
  # @returns JSON Object with deleted_id
  # @throws JSON Object of Errors
  def destroy 
    post = Post.find(params[:id])
    if post.destroy
      render json: { deleted_id: params[:id].to_i }
    else 
      render json: post.errors
    end 
  end 

  private

    def post_params
      # Rails escapes by default, so we only permit those.
      params.require(:post).permit(:title, :body, :user_id, :url_path)
    end

    # Make sure post exists.  That way if not, we can throw a 404
    def ensure_post_exists
      begin
        Post.find(params[:id])
      rescue
        render_not_found('post', (params[:id]))
      end
    end   

    # Make sure user exists.  That way if not, we can throw a 404
    def ensure_user_exists
      begin
        User.find(params[:user_id])
      rescue
        render_not_found('user', (params[:user_id]))
      end
    end    

    def check_url_path
      if params[:post].has_key?(:url_path) 
        @url_path = (params[:post][:url_path]).to_s.parameterize
      else  
        @url_path = (params[:post][:title]).to_s.parameterize + '-' + Time.now.utc.iso8601
      end
    end

end
